server/src/main/java/org/elasticsearch/timeseries/support/TimeSeriesMetrics.java

```diff
@@ -91,262 +120,409 @@
         private final MetricsCallback callback;
-        private final TemporalAccessor from;
-        private final TemporalAccessor to;
         private final SearchRequest search;
-        private Map<String, Object> previousDimensions;
 
-        ValuesInRangeResponseHandler(
-            String metric,
-            MetricsCallback callback,
-            TemporalAccessor from,
-            TemporalAccessor to,
-            SearchRequest search,
-            @Nullable Map<String, Object> previousDimensions
-        ) {
-            this.metric = metric;
+        SearchResponseHandler(List<String> resolvedMetrics, MetricsCallback callback, SearchRequest search) {
+            this.resolvedMetrics = resolvedMetrics;
+            this.previousDimensions = null;
             this.callback = callback;
-            this.from = from;
-            this.to = to;
             this.search = search;
-            this.previousDimensions = previousDimensions;
         }
 
         @Override
         public void accept(SearchResponse response) {
             // TODO shard error handling
-            logger.debug("Received batch of values {} with {} docs", search, response.getHits().getHits().length);
             SearchHit[] hits = response.getHits().getHits();
-            for (SearchHit hit : hits) {
-                /*
-                 * Read the dimensions out of the sort. This is useful because
-                 * we already need the sort so we can do proper pagination but
-                 * it also converts numeric dimension into a Long which is nice
-                 * and consistent.
-                 */
-                Map<String, Object> dimensions = new HashMap<>();
-                for (int d = 0; d < dimensionFieldNames.size(); d++) {
-                    Object dimensionValue = hit.getSortValues()[d];
-                    if (dimensionValue != null) {
-                        dimensions.put(dimensionFieldNames.get(d), dimensionValue);
+            for (String metric : resolvedMetrics) {
+                for (SearchHit hit : hits) {
+                    /*
+                     * Read the dimensions out of the sort. This is useful because
+                     * we already need the sort so we can do proper pagination but
+                     * it also converts numeric dimension into a Long which is nice
+                     * and consistent.
+                     */
+                    Map<String, Object> dimensions = new HashMap<>();
+                    for (int d = 0; d < dimensionFieldNames.size(); d++) {
+                        Object dimensionValue = hit.getSortValues()[d];
+                        if (dimensionValue != null) {
+                            dimensions.put(dimensionFieldNames.get(d), dimensionValue);
+                        }
                     }
+                    DocumentField metricField = hit.field(metric);
+                    if (metricField == null) {
+                        // TODO skip in query?
```

Comment:
This has to stay forever I think. You are filtering out documents that don't have any of the fields, but we still have to skip those that have only one.